Make build cache ignore mtime

jlhawn opened the pull request:
Build cache uses pgk/tarsum to get a digest of content which is
ADD'd or COPY'd during a build. The builder has always used v0 of
the tarsum algorithm which includes mtimes however since the whole
file is hashed anyway, the mtime doesn't really provide any extra
information about whether the file has changed and many version
control tools like Git strip mtime from files when they are cloned.

This patch updates the build subsystem to use v1 of Tarsum which
explicitly ignores mtime when calculating a digest. Now ADD and
COPY will result in a cache hit if only the mtime and not the file
contents have changed.

NOTE: Tarsum is NOT a meant to be a cryptographically secure hash
function. It is a best-effort approach to determining if two sets of
filesystem content are different.

Commit:
[builder] Make build cache ignore mtime

Build cache uses pgk/tarsum to get a digest of content which is
ADD'd or COPY'd during a build. The builder has always used v0 of
the tarsum algorithm which includes mtimes however since the whole
file is hashed anyway, the mtime doesn't really provide any extra
information about whether the file has changed and many version
control tools like Git strip mtime from files when they are cloned.

This patch updates the build subsystem to use v1 of Tarsum which
explicitly ignores mtime when calculating a digest. Now ADD and
COPY will result in a cache hit if only the mtime and not the file
contents have changed.

NOTE: Tarsum is NOT a meant to be a cryptographically secure hash
function. It is a best-effort approach to determining if two sets of
filesystem content are different.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
@jfrazelle @icecrime I'm not sure if someone has already submitted a patch like this. It seems like an optional flag to ADD/COPY is also a popular idea: https://github.com/docker/docker/issues/4351

duglin commented:
ping @erikh @tiborvass 
I'm worried about breaking backwards compat on this.  So, perhaps we could control this feature via a cmd flag on ADD/COPY once #10775 is merged.

jlhawn commented:
@duglin 

> I'm worried about breaking backwards compat on this.

By backwards compat do you mean how it would result in a cache miss the first time people use it?

> So, perhaps we could control this feature via a cmd flag on ADD/COPY

I'm agreeable to that. I can update this PR with some conditionals for using tarsum.v0 or tarsum.v1

cpuguy83 commented:
I know @tianon has given an explanation as to why just disabling it wouldn't be good, but I can never remember the reason.

I don't know a single person that likes the fact that mtime is used.

duglin commented:
@jlhawn yup - just worried about people who might count on today's behavior.

duglin commented:
e.g. haven't had to do this in a while but I know in a past life I would just "touch ..." something to force a rebuild of it - so a content-only check would break those cases.

cpuguy83 commented:
I guess you can currently do something like `touch cachebreaker` to deliberatly break the cache

jlhawn commented:
@duglin couldn't you just use `--no-cache`? Though it may be a sledge hammer for a small nail if you just wanted to invalidate one cached layer.

duglin commented:
@jlhawn yep exactly, its a bit of an overkill hammer.  Kind of reminds me of the discussion in #10682

phemmer commented:
Instead of a flag on ADD/COPY, what about a flag on `docker build`? So far none of the Dockerfile commands have flags. While I'm not opposed to them, I feel like they shouldn't be added unless needed.

To me, a flag on `docker build` even makes more sense. If you're working in a repo (build context) that doesn't preserve mtimes, it's going to be mtimes for every file in the repo that isn't trustworthy, not just one or two files.

icecrime commented:
I like the proposed solution, I'm not sure I see the point of an extra flag for this.

However, this breaks some tests:

```
--- FAIL: TestBuildADDCurrentDirWithCache (2.04s)
	docker_cli_build_test.go:3033: The cache should have been invalided but hasn't.
--- FAIL: TestBuildADDRemoteFileMTime (2.59s)
	docker_cli_build_test.go:3201: The cache should not have been used but was
```

duglin commented:
@phemmer a flag on `docker build` is possible as long as people are comfortable with it applying to all ADD/COPY commands and don't need to be selective.  Having said that though, I worry about forcing the person issuing the `docker build` to remember to include this flag each time. I suspect that many times the author of the Dockerfile knows about the issue and will want some way to control it from within the Dockerfile to ensure that all users of the Dockerfile get the same caching behavior. 
Of course having to specify it on each ADD/COPY is annoying (In practice I do wonder whether it would really be needed on all or just a small handful).  In some other previous issue/PR, the idea of setting global flags from within the Dockerfile was mentioned, perhaps this might be a good usecase for it?

icecrime commented:
Ping @jlhawn: can you please fix the tests?

duglin commented:
@icecrime can we get https://github.com/docker/docker/pull/10775 behind us then we can explore all of the options available for features like this mtime one?  I'm not sure we should change the current default behavior, so I think leveraging CMD flags might be a better choice.  So, I'm not sure there's a point in tweaking code in PRs like this *if* we may go a different direction after #10775

stevvooe commented:
:+1: for the design. The build cache should use the latest version of tarsum. The debate over whether or not mtime should apply belongs in a tarsum issue.

LK4D4 commented:
ping @jlhawn looks like tests still failing

jlhawn commented:
@LK4D4 @icecrime @duglin sorry, I forgot about this PR. Do we want to go with this solution? or should I close it in favor of  the `cache-strategy` flag approach?

LK4D4 commented:
I'd go with this.
ping @docker/core-maintainers

duglin commented:
I don't see how we can go with this since it changes the behavior people will see.  For example, if I touch a file in my build context to force a cache miss on a COPY/ADD, this PR will break me.  I would prefer if we offered a flag on ADD/COPY to allow people to opt-in to the new behavior.

phemmer commented:
> it changes the behavior people will see

Does docker have a compatibility guarantee between releases? Are we locked into this behavior until docker 2.0?

duglin commented:
Don't know if its written down but its generally good practice not to change semantics on people in minor (.x) releases.

tiborvass commented:
@phemmer if we decide we want this, then it would be in a non-minor Docker release (next one being 1.7).

jlhawn commented:
@tiborvass semantic version numbers are `<major>.<minor>.<patch>`, so Docker 1.7.0 is a *minor* version change from Docker 1.6.1

tiborvass commented:
@jlhawn you are correct, I edited it.

LK4D4 commented:
I definitely prefer current behavior over flags to instructions. But I prefer `tarsum.V2` over both.

stevvooe commented:
Note that tarsum v0 also doesn't take into account xattrs, so this could be considered a bug.

thaJeztah commented:
> if I touch a file in my build context to force a cache miss on a COPY/ADD

Well, you could still `date >> foobar` in stead of `touch`. However, IMO the cache should only break if the build-context is actually different. I don't think `mtime` should be taken into account. Apart from "manually" breaking the cache, do we know an **actual** use-case where the mtime *should* break the case (i.e., is important for the resulting image). If there are cases, I think they are very exceptional. We could always add an `--use-mtime` option :)

> Note that tarsum v0 also doesn't take into account xattrs, so this could be considered a bug.

Wasn't there an issue with AUFS not supporting xattrs? Or does my memory fail me here?

duglin commented:
I think you’re missing the point.  Its not about whether someone could change their logic to work with mtime not being a factor, its whether they are impacted at all. Today, we know that mtime is taken into account, for better or worse, and changing that is a pretty significant change IMO and changing our default behavior like this should not be done in a dot-release.  If we could convince ourselves its a bug then that would be different, but I’m not sure this would qualify as a bug since it was done on purpose.

-Doug

> On May 12, 2015, at 7:12 PM, Sebastiaan van Stijn <notifications@github.com> wrote:
> 
> if I touch a file in my build context to force a cache miss on a COPY/ADD
> 
> Well, you could still date >> foobar in stead of touch. However, IMO the cache should only break if the build-context is actually different. I don't think mtime should be taken into account. Apart from "manually" breaking the cache, do we know an actual use-case where the mtime should break the case (i.e., is important for the resulting image). If there are cases, I think they are very exceptional. We could always add an --use-mtime option :)
> 
> Note that tarsum v0 also doesn't take into account xattrs, so this could be considered a bug.
> 
> Wasn't there an issue with AUFS not supporting xattrs? Or does my memory fail me here?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/docker/docker/pull/12031#issuecomment-101483352>.
>

thaJeztah commented:
I agree with not causing BC breaking changes (https://github.com/docker/docker/issues/4907), but wonder if changing the cache behavior is considered a well-described feature, or just an implementation detail (i.e., *"Docker offers a caching mechanism to try and speed up the build process where possible"*). The resulting image will still be the same.

I think the actual caching is not in a "specs" somewhere and only described here; https://docs.docker.com/articles/dockerfile_best-practices/#build-cache

> In the case of the ADD and COPY instructions, the contents of the file(s) being put into the image are examined. Specifically, a checksum is done of the file(s) and then that checksum is used during the cache lookup. If anything has changed in the file(s), including its metadata, then the cache is invalidated.

Perhaps a `--build-opts` daemon-level flag; (or won't that work for builder?)

stevvooe commented:
I don't see any reason to hold the platform back on this account. Tarsum V0 doesn't correctly handle xattr or mtime, which is part of the contract of tarsum. Taking mtime into account for the build cache is not the contract of the build cache (ie using `touch` "because i tried it and it works"  is not a contract). Adding more flags makes the problem even worse. The fact that fixing this bug changes behavior is a moot point. It was broken before and this fixes it.

Let's get this merged.

jlhawn commented:
I agree with @stevvooe's last comment, however the tests that are failing are failing because they expect this behavior where mtime matters, etc.

duglin commented:
@jlhawn exactly.  Its been my understanding that the mtime part of the check was done on purpose and not a bug.  But I don't have the full history - this is just based on what I've heard about how conservative we are w.r.t. finding hits in the cache.  For example, when we download a file we don't use the last-modified http header to determine if its been modified, we download it and check the contents to be extra sure.
/ping @erikh

jlhawn commented:
Though I guess one could argue that these tests aren't really testing any *documented* behavior.

cpuguy83 commented:
In everything I've read changes here have always been centered around removing mtime from the caching mechanism but it always affected other things, such as losing the other metadata like file perms... I don't think anyone has said that we really need mtime here (and am willing to be proven wrong here).

I don't think it makes sense other than tricking the builder into not caching at a particular layer... and there are other ways to handle that.

stevvooe commented:
@cpuguy83 :+1: 

Taking _mtime_ into account should be considered a bug.

LK4D4 commented:
@jlhawn Sorry for long responses. Would you mind to fix tests?

jlhawn commented:
will do :+1:

jlhawn commented:
@duglin @LK4D4 @cpuguy83 @icecrime I updated this yesterday to modify the tests to accept the updated build cache behavior, but it's been a cat-and-mouse game with Jenkins ever since while some non-related and non-deterministic integration tests keep failing :-(

jlhawn commented:
@duglin @LK4D4 @cpuguy83 @icecrime IT WORKD! YAY!

LK4D4 commented:
LGTM

icecrime commented:
+1 that taking mtime into account is a bug. I've seen [numerous](http://tech.bam-x.com/blog/2015/03/31/docker-cache-is-busted/) [blog](http://blog.atish.me/2015/05/14/docker-cache-insights/) posts explaining how to force mtime, or reset it to the last commit time after a git clone, to workaround this issue.

LGTM

icecrime commented:
So, what about actually documenting this now? ;-)

thaJeztah commented:
> So, what about actually documenting this now? ;-)

Yeah! Me loves docs!

To start with, I think there's a section on how to work with and break the cache, that will need an update.

Typing on my phone now, but ping me if you can't find it :)

duglin commented on integration-cli/docker_cli_build_test.go:
```diff
@@ -2951,8 +2943,8 @@ func (s *DockerSuite) TestBuildAddRemoteFileMTime(c *check.C) {
 
-	// Now create a different server withsame contents (causes different mtim)
-	// This time the cache should not be used
+	// Now create a different server withsame contents (causes different mtime)
```
very minor (and it was there before), but:  s/withsame/with same/

duglin commented:
LGTM

ping @crosbymichael @shykes @erikh due to the UX impact/backwards compat issue

thaJeztah commented:
@jlhawn a quick search; 

this part of the docs is explaining the cache;
http://docs.docker.com/articles/dockerfile_best-practices/#build-cache

`ADD` mentions mtime as being included in invalidating the cache; http://docs.docker.com/reference/builder/#add

stevvooe commented:
@thaJeztah The use of `mtime` is a side-effect of a bug in tarsum. The fact that this has leaked into the `ADD` documentation shouldn't matter. The use of build cache with `mtime` makes no sense as a feature.

It makes no sense that this discussion is still ongoing. This fix should have been a part of 1.7. Let's stop holding the project back on broken minutiae.

thaJeztah commented:
@stevvooe sorry if I gave the impression I was against this, I'm 100% for this change; this was a major PITA in docker.

I just searched the docs, to assist @jlhawn pointing out where changes are needed, and this PR can be completed with those changes and get merged.

I'd be +1 if there's a chance to still get this in the 1.7 release (if we treat it as a bugfix and not a feature). If that means having the docs-changes in a follow-up PR (they're really minimal), I'm fine with that.

Sorry for the confusion.
--

> On 4 jun. 2015, at 23:17, Stephen Day <notifications@github.com> wrote:
> 
> @thaJeztah The use of mtime is a side-effect of a bug in tarsum. The fact that this has leaked into the ADD documentation shouldn't matter. The use of build cache with mtime makes no sense as a feature.
> 
> It makes no sense that this discussion is still ongoing. This fix should have been a part of 1.7. Let's stop holding the project back on broken minutiae.
> 
> —
> Reply to this email directly or view it on GitHub.
>

stevvooe commented:
@thaJeztah No worries. I appreciate your attention to detail. That statement was for everyone but came off harsh. Really, this change shouldn't take 3 months.

LK4D4 commented:
Yeah, we have enough LGTMs for code, now docs time :)

Commit:
[docs] Update builder docs on last-modified times

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
@LK4D4 @thaJeztah docs have been updated. Please take a look.

thaJeztah commented:
LGTM, thanks @jlhawn !

ping @fredlf @jamtur01 @moxiegirl @SvenDowideit

jlhawn commented:
oh, I should go back and correct the typo pointed out by @duglin too. one sec...

moxiegirl commented:
LGTM

cpuguy83 commented:
![](http://i.imgur.com/fIhLNdo.gif)

thaJeztah commented:
\o/ don't know **when** it happened, lost track of mtime :smile:

srcspider commented:
What release is this scheduled for?
